### PR TITLE
Clean up webpack config, increase build speed

### PR DESCRIPTION
**Pull request by timneutkens**
This is a first step towards more comments in the core explaining what/why thing are happening.
As a side effect I dug deep into uglify and why it's so slow when building Next.

I did some small benchmarks on the 3 versions: master, canary, this pr. Here's the results. Others compared to master using a large app:

4.2.1: 78.50s **0%**
4.3.0-canary.1: 79.90s **+1.78%**
New: 50.66s **-35%**

### CODE REVIEW

File: server/build/plugins/combine-assets-plugin.js [hunk 1]
```diff
@@ -11,26 +11,19 @@ export default class CombineAssetsPlugin {
 
   apply (compiler) {
     compiler.plugin('compilation', (compilation) => {
-      compilation.plugin('additional-chunk-assets', (chunks) => {
```
Comment (arunoda) [inline server/build/plugins/combine-assets-plugin.js]: I think we've this as a separate PR?
So, you want to close that or remove these changes from here?

File: server/config.js [hunk 2]
```diff
@@ -9,8 +9,7 @@ const defaultConfig = {
   distDir: '.next',
   assetPrefix: '',
   configOrigin: 'default',
-  useFileSystemPublicRoutes: true,
-  pagesGlobPattern: 'pages/**/*.+(js|jsx)'
```
Comment (arunoda) [inline server/config.js]: Why we've remove this?

File: server/build/webpack.js [hunk 3]
```diff
@@ -356,7 +275,123 @@ export default async function createCompiler (dir, { buildId, dev = false, quiet
...
+
+                if (existsSync(filePath)) {
+                  throw new Error(`Both ${from} and ${to} file found. Please make sure you only have one of both.`)
+                }
+              }
+            },
+            // By default, our babel config does not transpile ES2015 module syntax because
+            // webpack knows how to handle them. (That's how it can do tree-shaking)
+            // But Node.js doesn't know how to handle them. So, we have to transpile them here.
+            transform ({ content, sourceMap, interpolatedName }) {
```
Comment (arunoda) [inline server/build/webpack.js]: Move this content into a function.

File: server/build/webpack.js [hunk 4]
```diff
@@ -356,7 +275,123 @@ export default async function createCompiler (dir, { buildId, dev = false, quiet
     },
     plugins,
     module: {
-      rules
+      rules: [
+        ...devLoaders,
+        {
```
Comment (arunoda) [inline server/build/webpack.js]: May be we can have other loaders defined above just like `devLoaders`

**Review by arunoda — CHANGES_REQUESTED** [review]

File: server/build/webpack.js (hunk 4 above)
Comment (timneutkens) [reply server/build/webpack.js]: 👍

File: server/build/plugins/combine-assets-plugin.js (hunk 1 above)
Comment (timneutkens) [reply server/build/plugins/combine-assets-plugin.js]: Merged 👍

File: server/config.js (hunk 2 above)
Comment (timneutkens) [reply server/config.js]: Re-added.

**Comment by arunoda** [comment]
@timneutkens this looks great to me.
Is this mergeable?

**Comment by timneutkens** [comment]
@arunoda, yes!

**Pull request merged**